File: modules/printer-helper/android/src/main/java/church/b1/checkin/printer/PrinterHelperModule.kt
package church.b1.checkin.printer

import android.graphics.Bitmap
import android.net.Uri
import android.provider.MediaStore
import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

class PrinterHelperModule : Module() {
    private val providers: Map<String, PrintProviderInterface> = mapOf(
        "Brother" to BrotherProvider(),
        "Zebra" to ZebraProvider()
    )
    private var activeProvider: PrintProviderInterface? = null
    private var status = "Pending init"

    override fun definition() = ModuleDefinition {
        Name("PrinterHelper")

        Events("StatusUpdated", "onError", "onEvent")

        // Wire up callbacks for all providers
        for ((_, provider) in providers) {
            provider.onError = { source, message ->
                sendEvent("onError", bundleOf("source" to source, "message" to message))
            }
            provider.onEvent = { eventType, source, message ->
                sendEvent("onEvent", bundleOf("eventType" to eventType, "source" to source, "message" to message))
            }
        }

        AsyncFunction("scan") {
            val executor = Executors.newFixedThreadPool(providers.size)
            val futures = providers.map { (brand, provider) ->
                brand to executor.submit<Array<String>> { provider.scan() }
            }

            val allResults = mutableListOf<String>()
            for ((brand, future) in futures) {
                try {
                    val results = future.get()
                    for (result in results) {
                        allResults.add("$brand~$result")
                    }
                } catch (_: Exception) {
                    // Skip providers that fail to scan
                }
            }
            executor.shutdown()

            allResults.joinToString(",")
        }

        Function("checkInit") { ip: String, model: String, brand: String ->
            val context = appContext.reactContext ?: return@Function
            val providerBrand = brand.ifEmpty { "Brother" }
            activeProvider = providers[providerBrand] ?: providers["Brother"]
            activeProvider?.checkInit(context, ip, model)
            if (ip.isNotEmpty()) {
                updateStatus(ip)
            } else {
                updateStatus("No Printer")
            }
        }

        Function("printUris") { uriList: String ->
            val context = appContext.reactContext ?: return@Function
            val provider = activeProvider ?: return@Function
            val uris = uriList.split(",")
            val bmps = mutableListOf<Bitmap>()

            for (uriString in uris) {
                try {
                    val uri = Uri.parse(uriString)
                    @Suppress("DEPRECATION")
                    val bitmap = MediaStore.Images.Media.getBitmap(context.contentResolver, uri)
                    bmps.add(bitmap)
                } catch (_: Exception) {
                    // Skip invalid URIs
                }
            }

            if (bmps.isNotEmpty()) {
                provider.printBitmaps(bmps)
            }
        }

        Function("configure") {
            activeProvider?.configure()
        }

        Function("getStatus") {
            status
        }
    }

    private fun updateStatus(newStatus: String) {
        status = newStatus
        sendEvent("StatusUpdated", bundleOf("status" to status))
    }
}

File: modules/printer-helper/ios/src/PrinterHelperModule.swift
import ExpoModulesCore
import BRLMPrinterKit

public class PrinterHelperModule: Module {
    private var providers: [String: PrintProviderProtocol] = [:]
    private var activeProvider: PrintProviderProtocol?

    private func setupProviders() {
        var brother = BrotherProvider()
        brother.onError = { [weak self] source, message in
            self?.sendEvent("onError", ["source": source, "message": message])
        }
        brother.onEvent = { [weak self] eventType, source, message in
            self?.sendEvent("onEvent", ["eventType": eventType, "source": source, "message": message])
        }
        providers["Brother"] = brother

        var zebra = ZebraProvider()
        zebra.onError = { [weak self] source, message in
            self?.sendEvent("onError", ["source": source, "message": message])
        }
        zebra.onEvent = { [weak self] eventType, source, message in
            self?.sendEvent("onEvent", ["eventType": eventType, "source": source, "message": message])
        }
        providers["Zebra"] = zebra
    }

    public func definition() -> ModuleDefinition {
        Name("PrinterHelper")

        Events("StatusUpdated", "onError", "onEvent")

        OnCreate {
            self.setupProviders()
        }

        AsyncFunction("scan") { () -> String in
            var allResults: [String] = []
            let group = DispatchGroup()
            let lock = NSLock()

            for (brand, provider) in self.providers {
                group.enter()
                DispatchQueue.global().async {
                    let results = provider.scan()
                    let prefixed = results.map { "\(brand)~\($0)" }
                    lock.lock()
                    allResults.append(contentsOf: prefixed)
                    lock.unlock()
                    group.leave()
                }
            }

            group.wait()
            return allResults.joined(separator: ",")
        }

        Function("checkInit") { (ip: String, model: String, brand: String) in
            let providerBrand = brand.isEmpty ? "Brother" : brand
            self.activeProvider = self.providers[providerBrand] ?? self.providers["Brother"]
            self.activeProvider?.checkInit(ip: ip, model: model)
            if let status = self.activeProvider?.status {
                self.sendEvent("StatusUpdated", ["status": status])
            }
        }

        Function("printUris") { (uriList: String) in
            let uris = uriList.components(separatedBy: ",").filter { !$0.isEmpty }
            self.activeProvider?.printImages(uris)
        }

        Function("configure") {
            self.activeProvider?.configure()
        }

        Function("getStatus") { () -> String in
            return self.activeProvider?.status ?? "No Printer"
        }
    }
}
